File: src/collectors/filters/diversitySelector.ts
import type { Article, NewsCategory } from "../../domain/article.js";
import type { RelevanceScore } from "./relevanceScorer.js";
import { DIVERSITY } from "../../config/constants.js";

export interface DiversitySelectionOptions {
  maxArticlesPerSource?: number;
  maxArticlesPerCategory?: number;
  minRelevanceScore?: number;
  maxTotal?: number;
}

export interface DiversitySelectionResult {
  selected: Article[];
  excluded: Article[];
  stats: {
    bySource: Record<string, number>;
    byCategory: Record<string, number>;
    excludedBySourceLimit: number;
    excludedByCategoryLimit: number;
    excludedByRelevance: number;
  };
}

/**
 * Selects articles considering source and category diversity.
 *
 * Selection priority:
 * 1. Relevance score (higher = better)
 * 2. Source diversity (cap per source)
 * 3. Category diversity (cap per category)
 *
 * Quality and relevance are prioritized; diversity acts as a tiebreaker
 * and prevents monopolization by a single source or category.
 */
export function selectWithDiversity(
  articles: Article[],
  scores: RelevanceScore[],
  options: DiversitySelectionOptions = {},
): DiversitySelectionResult {
  const maxPerSource = options.maxArticlesPerSource ?? DIVERSITY.MAX_ARTICLES_PER_SOURCE;
  const maxPerCategory = options.maxArticlesPerCategory ?? DIVERSITY.MAX_ARTICLES_PER_CATEGORY;
  const minRelevance = options.minRelevanceScore ?? DIVERSITY.MIN_PERSONAL_FINANCE_RELEVANCE;
  const maxTotal = options.maxTotal ?? Infinity;

  const scoreMap = new Map<string, number>();
  for (const s of scores) {
    scoreMap.set(s.articleId, s.score);
  }

  // Sort by relevance score (descending), then by published time (descending for freshness)
  const sorted = [...articles].sort((a, b) => {
    const scoreA = scoreMap.get(a.id) ?? 0;
    const scoreB = scoreMap.get(b.id) ?? 0;
    if (scoreB !== scoreA) return scoreB - scoreA;
    return new Date(b.publishedAt).getTime() - new Date(a.publishedAt).getTime();
  });

  const selected: Article[] = [];
  const excluded: Article[] = [];
  const sourceCounts: Record<string, number> = {};
  const categoryCounts: Record<string, number> = {};

  let excludedBySourceLimit = 0;
  let excludedByCategoryLimit = 0;
  let excludedByRelevance = 0;

  for (const article of sorted) {
    if (selected.length >= maxTotal) {
      excluded.push(article);
      continue;
    }

    const relevance = scoreMap.get(article.id) ?? 0;

    // Filter by minimum relevance
    if (relevance < minRelevance) {
      excluded.push(article);
      excludedByRelevance++;
      continue;
    }

    // Check source limit
    const sourceCount = sourceCounts[article.sourceName] ?? 0;
    if (sourceCount >= maxPerSource) {
      excluded.push(article);
      excludedBySourceLimit++;
      continue;
    }

    // Check category limit - article passes if at least one category has room
    const articleCategories = article.categories.length > 0 ? article.categories : ["other" as NewsCategory];
    const allCategoriesFull = articleCategories.every(
      (cat) => (categoryCounts[cat] ?? 0) >= maxPerCategory,
    );

    if (allCategoriesFull) {
      excluded.push(article);
      excludedByCategoryLimit++;
      continue;
    }

    // Accept article
    selected.push(article);
    sourceCounts[article.sourceName] = sourceCount + 1;
    for (const cat of articleCategories) {
      categoryCounts[cat] = (categoryCounts[cat] ?? 0) + 1;
    }
  }

  return {
    selected,
    excluded,
    stats: {
      bySource: { ...sourceCounts },
      byCategory: { ...categoryCounts },
      excludedBySourceLimit,
      excludedByCategoryLimit,
      excludedByRelevance,
    },
  };
}

File: src/main/java/com/economicbriefing/collector/filter/DiversitySelector.java
package com.economicbriefing.collector.filter;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.economicbriefing.config.AppProperties;
import com.economicbriefing.domain.article.Article;
import com.economicbriefing.domain.article.NewsCategory;
import org.springframework.stereotype.Component;

@Component
public class DiversitySelector {

    private final AppProperties appProperties;

    public DiversitySelector(AppProperties appProperties) {
        this.appProperties = appProperties;
    }

    public DiversitySelectionResult select(
            List<Article> articles,
            List<RelevanceScorer.RelevanceScore> scores,
            DiversityOptions options
    ) {
        int maxPerSource = options.maxArticlesPerSource() != null
                ? options.maxArticlesPerSource()
                : appProperties.diversity().maxArticlesPerSource();
        int maxPerCategory = options.maxArticlesPerCategory() != null
                ? options.maxArticlesPerCategory()
                : appProperties.diversity().maxArticlesPerCategory();
        int minRelevance = options.minRelevanceScore() != null
                ? options.minRelevanceScore()
                : appProperties.diversity().minPersonalFinanceRelevance();
        int maxTotal = options.maxTotal() != null ? options.maxTotal() : Integer.MAX_VALUE;
        int softMaxOverride = options.softMaxOverrideScore() != null
                ? options.softMaxOverrideScore()
                : appProperties.diversity().softMaxOverrideScore();

        Map<String, Integer> scoreMap = new HashMap<>();
        for (RelevanceScorer.RelevanceScore s : scores) {
            scoreMap.put(s.articleId(), s.score());
        }

        // Sort by relevance score (descending), then by published time (descending)
        List<Article> sorted = new ArrayList<>(articles);
        sorted.sort(Comparator
                .<Article, Integer>comparing(a -> scoreMap.getOrDefault(a.id(), 0), Comparator.reverseOrder())
                .thenComparing(Article::publishedAt, Comparator.reverseOrder()));

        List<Article> selected = new ArrayList<>();
        List<Article> excluded = new ArrayList<>();
        Map<String, Integer> sourceCounts = new HashMap<>();
        Map<String, Integer> categoryCounts = new HashMap<>();

        int excludedBySourceLimit = 0;
        int excludedByCategoryLimit = 0;
        int excludedByRelevance = 0;

        for (Article article : sorted) {
            if (selected.size() >= maxTotal) {
                excluded.add(article);
                continue;
            }

            int relevance = scoreMap.getOrDefault(article.id(), 0);

            if (relevance < minRelevance) {
                excluded.add(article);
                excludedByRelevance++;
                continue;
            }

            // Check source limit (hard cap)
            int sourceCount = sourceCounts.getOrDefault(article.sourceName(), 0);
            if (sourceCount >= maxPerSource) {
                excluded.add(article);
                excludedBySourceLimit++;
                continue;
            }

            // Check category limit (soft cap)
            List<NewsCategory> articleCategories = (article.categories() != null && !article.categories().isEmpty())
                    ? article.categories()
                    : List.of(NewsCategory.OTHER);

            boolean allCategoriesFull = articleCategories.stream().allMatch(cat -> {
                int count = categoryCounts.getOrDefault(cat.toValue(), 0);
                int effectiveMax = relevance >= softMaxOverride ? maxPerCategory + 1 : maxPerCategory;
                return count >= effectiveMax;
            });

            if (allCategoriesFull) {
                excluded.add(article);
                excludedByCategoryLimit++;
                continue;
            }

            selected.add(article);
            sourceCounts.merge(article.sourceName(), 1, Integer::sum);
            for (NewsCategory cat : articleCategories) {
                categoryCounts.merge(cat.toValue(), 1, Integer::sum);
            }
        }

        DiversityStats stats = new DiversityStats(
                new HashMap<>(sourceCounts), new HashMap<>(categoryCounts),
                excludedBySourceLimit, excludedByCategoryLimit, excludedByRelevance
        );

        return new DiversitySelectionResult(selected, excluded, stats);
    }

    public record DiversityOptions(
            Integer maxArticlesPerSource,
            Integer maxArticlesPerCategory,
            Integer minRelevanceScore,
            Integer maxTotal,
            Integer softMaxOverrideScore
    ) {
        public static DiversityOptions defaults() {
            return new DiversityOptions(null, null, null, null, null);
        }
    }

    public record DiversitySelectionResult(
            List<Article> selected,
            List<Article> excluded,
            DiversityStats stats
    ) {}

    public record DiversityStats(
            Map<String, Integer> bySource,
            Map<String, Integer> byCategory,
            int excludedBySourceLimit,
            int excludedByCategoryLimit,
            int excludedByRelevance
    ) {}
}
